ws-server/src/future.rs

```diff
@@ -130,3 +136,72 @@ where
 		this.selector.poll_unpin(cx)
 	}
 }
+
+#[derive(Debug)]
+struct MonitorInner {
+	shutdown_requested: AtomicBool,
+	waker: AtomicWaker,
+}
+
+/// Monitor for checking whether the server has been flagged to shut down.
+#[derive(Debug, Clone)]
+pub(crate) struct StopMonitor(Arc<MonitorInner>);
+
+impl Drop for StopMonitor {
+	fn drop(&mut self) {
+		if Arc::strong_count(&self.0) == 1 {
+			self.0.waker.wake();
+		}
+	}
+}
+
+impl StopMonitor {
+	pub(crate) fn new() -> Self {
+		StopMonitor(Arc::new(MonitorInner { shutdown_requested: AtomicBool::new(false), waker: AtomicWaker::new() }))
+	}
+
+	pub(crate) fn shutdown_requested(&self) -> bool {
+		self.0.shutdown_requested.load(Ordering::Relaxed)
+	}
+
+	pub(crate) fn handle(&self) -> StopHandle {
+		StopHandle(Arc::downgrade(&self.0))
+	}
+}
+
+/// Handle that is able to stop the running server.
+#[derive(Debug, Clone)]
+pub struct StopHandle(Weak<MonitorInner>);
+
+impl StopHandle {
+	/// Requests server to stop. Returns an error if server was already stopped.
+	///
+	/// Returns a future that can be awaited for when the server shuts down.
+	pub fn stop(self) -> Result<ShutdownWaiter, Error> {
+		if let Some(arc) = Weak::upgrade(&self.0) {
```

Comment:
So, when `stop` is called, we add 1 to the `strong_count` of the arc (if it's still around).

So there exists a brief period of time I think in which, if the last `StopMonitor` is dropped, its `Drop` impl won't wake the waker because, at the same time, a `StopHandle` is holding another strong ref to it.

If that manages to be the case (a very slim chance of course) then you'll have no `StopMonitors` left, and the `ShutdownWaiter` future will wait forever for a waker that will never be woken, I think.

Reply:
I think it should be fine because the temporary arc is going to be dropped before the waiter is returned and gets a chance to be polled.

Reply:
Ah that's true!